### PR TITLE
AMDGPU: Move libcall simplify into PeepholeEP

### PR DESCRIPTION
**Pull request by arsenm**
We were running this immediately on the incoming IR, which
is still littered with temporary allocas obscuring trivial values.
This needs to run after initial SROA to handle sincos insertion.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/amdgpu-libcall-sincos-pass-ordering.ll [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  store double %conv, ptr addrspace(1) %out0, align 8
+  store double %conv5, ptr addrspace(1) %out1, align 8
+  store double %conv6, ptr addrspace(1) %out2, align 8
+  call void @llvm.lifetime.end.p5(i64 4, ptr addrspace(5) %z)
+  call void @llvm.lifetime.end.p5(i64 4, ptr addrspace(5) %y)
+  ret void
+}
+
+declare void @llvm.lifetime.start.p5(i64 immarg, ptr addrspace(5) nocapture) #1
+declare void @llvm.lifetime.end.p5(i64 immarg, ptr addrspace(5) nocapture) #1
```
Comment (Pierre-vh) [inline llvm/test/CodeGen/AMDGPU/amdgpu-libcall-sincos-pass-ordering.ll:60]: Do we still need intrinsic declarations?

File: llvm/test/CodeGen/AMDGPU/amdgpu-libcall-sincos-pass-ordering.ll [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+define internal float @_Z3sinf(float noundef %arg) #2 {
+bb:
+  %i = tail call float asm "pseudo-libcall-sin %0, %1", "=v,v"(float noundef %arg) #2
+  ret float %i
+}
+
+attributes #0 = { norecurse nounwind }
+attributes #1 = { nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) }
+attributes #2 = { mustprogress nofree norecurse nounwind willreturn memory(none) }
+attributes #3 = { nounwind willreturn memory(none) }
```
Comment (Pierre-vh) [inline llvm/test/CodeGen/AMDGPU/amdgpu-libcall-sincos-pass-ordering.ll:77]: Do we need all these attributes?

File: llvm/test/CodeGen/AMDGPU/amdgpu-libcall-sincos-pass-ordering.ll [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  %conv = fpext float %i4 to double
+  %i5 = load float, ptr addrspace(5) %y, align 4
+  %conv5 = fpext float %i5 to double
+  %i6 = load float, ptr addrspace(5) %z, align 4
+  %conv6 = fpext float %i6 to double
+  store double %conv, ptr addrspace(1) %out0, align 8
+  store double %conv5, ptr addrspace(1) %out1, align 8
+  store double %conv6, ptr addrspace(1) %out2, align 8
+  call void @llvm.lifetime.end.p5(i64 4, ptr addrspace(5) %z)
+  call void @llvm.lifetime.end.p5(i64 4, ptr addrspace(5) %y)
```
Comment (Pierre-vh) [inline llvm/test/CodeGen/AMDGPU/amdgpu-libcall-sincos-pass-ordering.ll:55]: are the lifetime intrinsics actually needed?

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/amdgpu-libcall-sincos-pass-ordering.ll:55]: They're more representative of actual IR the pass will see, which is part of the point

File: llvm/test/CodeGen/AMDGPU/amdgpu-libcall-sincos-pass-ordering.ll (hunk 2 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/amdgpu-libcall-sincos-pass-ordering.ll:77]: We need enough to get the asm calls deleted

File: llvm/test/CodeGen/AMDGPU/amdgpu-libcall-sincos-pass-ordering.ll (hunk 1 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/amdgpu-libcall-sincos-pass-ordering.ll:60]: Kind of no, but it's still canonically printed

**Review by rampitec — APPROVED** [review]

**Review by lamb-j — APPROVED** [review]
This fixes the issue I was having with my sincos() example. Thanks!

**Pull request merged**